src/protocols/bdx/AsyncTransferFacilitator.h

```diff
@@ -0,0 +1,122 @@
+public:
+    AsyncTransferFacilitator() : mExchange(*this) {}
+    ~AsyncTransferFacilitator() override;
+
+    /**
+     * This method should be implemented to contain business-logic handling of BDX messages
+     * and other TransferSession events.
+     *
+     * @param[in] event An OutputEvent that contains output from the TransferSession object.
+     */
+    virtual void HandleTransferSessionOutput(TransferSession::OutputEvent & event) = 0;
+
+    /**
+     * This method should be implemented to clean up and destroy the instance of the sub class
+     * implemeting it and the base class.
+     *
+     * Note: After this method has been called, we can't use the instance of the subclass that
+     * implements this.
+     */
+    virtual void CleanUp() = 0;
+
+protected:
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                 System::PacketBufferHandle && payload) override;
+    void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
+    void OnExchangeClosing(Messaging::ExchangeContext * ec) override;
+
+    CHIP_ERROR SendMessage(TransferSession::OutputEvent & event);
+
+    // The transfer session coresponding to this AsynTransferFacilitator object.
+    TransferSession mTransfer;
+
+    // The Exchange holder that holds the exchange context used for BDX messages.
+    Messaging::ExchangeHolder mExchange;
+
+    // The timeout for the BDX transfer session.
+    System::Clock::Timeout mTimeout;
+
+private:
+    // Releases the exchange context and calls reset on the BDX transfer session.
+    void Reset();
+};
+
+/**
+ * An AsyncTransferFacilitator that is initialized to respond to an incoming BDX transfer request.
+ *
+ * Provides a method for initializing the TransferSession member but still needs to be extended to implement
+ * HandleTransferSessionOutput.
+ *
+ * This class, or a subclass of it, should be used as the exchange delegate for a BDX transfer.
+ */
+class AsyncResponder : public AsyncTransferFacilitator
+{
+public:
+    /**
+     * Initialize the TransferSession state machine to be ready for an incoming transfer request.
+     *
+     * @param[in] exchangeCtx     The exchange to use for the transfer.
+     * @param[in] role            The role of the Responder: Sender or Receiver of BDX data
+     * @param[in] xferControlOpts Supported transfer modes (see TransferControlFlags)
+     * @param[in] maxBlockSize    The maximum supported size of BDX Block data
+     * @param[in] timeout         The chosen timeout delay for the BDX transfer
+     */
+    CHIP_ERROR PrepareForTransfer(Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                                  BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize,
+                                  System::Clock::Timeout timeout);
+
+    /**
+     * Notifies the transfer facilitator that an output event has been handled by the subclass that implements the
```

Comment:
Is it allowed to call this from inside `HandleTransferSessionOutput`?  I asked that question last time, and this is still not clearly documented.

Reply:
will add more documentation